### ISSUE TITLE
[FEATURE] run-script with workspaces should short-circuit on script error

### ISSUE THREAD

**Issue by johndiiorio** [npm/rfcs#575]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

`run-script` with the `--workspaces` flag will run all workspace scripts, regardless if one script exits with exit code 1. Please see the "Steps to Reproduce" section.

### Expected Behavior

I would expect that `run-script` would short-circuit and not run the other scripts if one fails. Alternatively, a new flag (e.g. `--short-circuit`) could be introduced to `run-script` in order to maintain backwards compatibility.

### Steps To Reproduce

Consider the following scenario:
- A root `package.json` file has `"workspaces": ["a", "b", "c"]`
- `a/package.json` has the script `"build": "exit 1"`
- Both `b/package.json` and `c/package.json` have the the script `"build": "exit 0"`

Run `npm run build --workspaces` and note that all scripts are run.

### Environment

- OS: Windows 10
- Node: 16.10.0
- npm: 7.24.0

**Comment by hapkecom** [comment]
Hi, is there any update on this? I thinks it's still needed for the reasons described above ...

**Comment by DavidSouther** [comment]
How would you expect npm to order script execution across packages?

**Comment by ljharb** [comment]
Topologically; in order of the dependency graph. Order would need to be deterministic for siblings but wouldn't otherwise matter.

**Comment by johnnyshankman** [comment]
Order already can be done deterministically @DavidSouther as i'm using workspaces in a repo where one pkg relies on the other, and therefore must be built first.

**Comment by johnnyshankman** [comment]
@darcyclarke any updates on this feature request? seems like a pretty easy one to add, just an exit check before moving on to the next workspace.

**Comment by ljharb** [comment]
@johnnyshankman darcy is no longer at npm, and npm hasn't done RFC calls in like a year, so there's not likely been any progress.

It would definitely not be easy to add; determining proper topological ordering is not a trivial task.

**Comment by johnnyshankman** [comment]
@ljharb thanks for the update.

i'm not so sure what you mean about the topologically ordering. meaning which workspace goes first?

**Comment by ljharb** [comment]
Yes, you might have 100 child workspaces, some of which have no workspace deps, but some of which are part of a complicated dependency chain. There may even be a circular dependency.

**Comment by johnnyshankman** [comment]
@ljharb how does it resolve the graph currently? why would we change that behavior? this would simply be an additional flag no?

**Comment by ljharb** [comment]
What I mean is, this feature only works properly in conjunction with topological sorting, and *that* is nontrivial to add.

**Comment by johnnyshankman** [comment]
@ljharb ah yes i see what you mean now. yes 100% in a perfect world we do topo sort and keep this on 24/7. agreed!

in the meantime, is it so bad to add a `--fail-fast` flag for `--workspaces` and those who have too complex of a repo for it to work would not use it?

my monorepo is one layer deep but 5 pkgs wide, so this would be hugely beneficial, as each subpkg relies on the last and building all of them in order every time to the end is silly when the first one breaks.

**Comment by johnnyshankman** [comment]
also curious if you know a modern workaround that avoids a feature request!

**Comment by darcyclarke** [comment]
> @darcyclarke any updates on this feature request? seems like a pretty easy one to add, just an exit check before moving on to the next workspace. is there any in-progress PRs to review etc?

I'm no longer working on npm unfortunately. That said, & fwiw, I do plan to ensure this is a capability out of the gate in my new project 🤷‍♂️

**Comment by krasevych** [comment]
Any update? I have the same problem

**Comment by EricVanDerDijsConnect** [comment]
Hi, is there any update regarding this?

**Comment by LupascuAndrei** [comment]
Would love to have an update on this

**Comment by zymotik** [comment]
https://github.com/npm/cli/pull/8323 has been sat waiting for a code review for over a year. This would be a very helpful feature to have!!

**Comment by ErwanRaulo** [comment]
> [@john-landgrave](https://github.com/john-landgrave) I agree, this bug makes it impractical to build a monorepo in a CI with just `npm`. One possible workaround is to create a script (and add it to your root `package.json`) that reads the `package.json` `workspaces` array, then individually invokes `npm run build --workspace <workspace>` and checks the exit code before proceeding. That shouldn't be necessary though, the behavior should be built-in.

that's exactly what it's done in my [npm package](https://github.com/ErwanRaulo/sumlyzer). it fails fast and run in parallèle workspaces if needed and other features. honest feedbacks are welcome.
